pkg/history/sqlitestore/store.go

```diff
@@ -0,0 +1,645 @@
+	return 0, nil
+}
+
+// builds an SQL term for filtering records based on source and ID range.
+// Also returns a slice of parameters to be passed when executing the query.
+// If no filtering is to be performed, returns a dummy condition to maintain valid SQL syntax.
+func buildFilters(source string, from, to RecordID) (string, []any) {
+	var filters []string
+	var args []any
+	if source != "" {
+		filters = []string{"history_sources.source = ?"}
+		args = []any{source}
+	}
+
+	if from != 0 {
+		filters = append(filters, "history.id >= ?")
+		args = append(args, from)
+	}
+
+	if to != 0 {
+		filters = append(filters, "history.id < ?")
+		args = append(args, to)
+	}
+
+	if len(filters) > 0 {
+		return strings.Join(filters, " AND "), args
+	} else {
+		return "1 = 1", nil // no filtering, return a dummy condition
+	}
+}
+
+type RecordID int64
+
+func MakeRecordID(ts time.Time, serial int) RecordID {
+	if serial < 0 || serial >= 1_000_000 {
+		panic("serial must be in range [0, 999999]")
+	}
+	return RecordID(ts.UnixMilli()*1_000_000 + int64(serial))
+}
+
+func ParseRecordID(s string) (RecordID, error) {
+	id, err := strconv.ParseInt(s, 16, 64)
+	if err != nil {
+		return 0, ErrInvalidRecordID
+	}
+	if id < 0 {
+		return 0, ErrInvalidRecordID
+	}
+	return RecordID(id), nil
+}
+
+func (id RecordID) Timestamp() time.Time {
+	return time.UnixMilli(int64(id) / 1_000_000)
+}
+
+func (id RecordID) Serial() int {
+	return int(int64(id) % 1_000_000)
+}
+
+func (id RecordID) Next() (RecordID, bool) {
+	if id.Serial() == 1_000_000-1 {
+		// incrementing the serial number would overflow the millisecond timestamp
+		return 0, false
+	}
+	return RecordID(int64(id) + 1), true
+}
+
+func (id RecordID) String() string {
+	return fmt.Sprintf("%016X", int64(id))
+}
+
+type sourceAllocator struct {
+	sources    map[string]int64 // source name to ID mapping
+	insertStmt *sql.Stmt        // prepared statement for inserting new sources if they don't exist
+	queryStmt  *sql.Stmt        // prepared statement for querying source IDs
+}
+
+func newSourceAllocator(ctx context.Context, tx *sql.Tx) (*sourceAllocator, error) {
+	sc := &sourceAllocator{
+		sources: make(map[string]int64),
+	}
+
+	var err error
+	sc.insertStmt, err = tx.PrepareContext(ctx, "INSERT OR IGNORE INTO history_sources (source) VALUES (?)")
+	if err != nil {
+		return nil, err
+	}
+
+	sc.queryStmt, err = tx.PrepareContext(ctx, "SELECT id FROM history_sources WHERE source = ?")
+	if err != nil {
+		return nil, err
+	}
+
+	return sc, nil
+}
+
+func (sa *sourceAllocator) getOrInsertSource(ctx context.Context, source string) (int64, error) {
+	if id, ok := sa.sources[source]; ok {
+		return id, nil
+	}
+
+	// Insert the source if it doesn't exist
+	_, err := sa.insertStmt.ExecContext(ctx, source)
+	if err != nil {
+		return 0, err
+	}
+
+	// Query the ID of the source
+	var id int64
+	err = sa.queryStmt.QueryRowContext(ctx, source).Scan(&id)
+	if err != nil {
+		return 0, err
+	}
```

Comment:
I'm curious why you've gone with `INSERT IGNORE` + `SELECT` instead of `SELECT` + `if absent { INSERT RETURNING ID }`. I feel like the latter is more idiomatic and potentially more performant as you could skip a statement.

Reply:
Changed the implementation as suggested.